chain/network/src/raw.rs

```diff
@@ -0,0 +1,345 @@
+                                lengths.push((l, timestamp));
+                                pos += l + 4;
+                                pending_msg_length = None;
+                            } else {
+                                break;
+                            }
+                        } else {
+                            break;
+                        }
+                    }
+                }
+                Err(e) => match e.kind() {
+                    io::ErrorKind::WouldBlock => {
+                        break;
+                    }
+                    _ => return Err(e),
+                },
+            };
+        }
+        Ok(lengths)
+    }
+
+    // Reads from the socket until there is at least one full PeerMessage available.
+    // After that point, continues to read any bytes available to be read without blocking
+    // and appends any extra messages to the returned Vec. Usually there is only one in there
+    async fn recv_messages(&mut self) -> io::Result<Vec<(PeerMessage, Instant)>> {
+        let mut messages = Vec::new();
+        let (msg_length, first_byte_time) = self.read_msg_length().await?;
+
+        while self.buf.remaining() < msg_length + 4 {
+            // TODO: measure time to last byte here
+            self.do_read().await?;
+        }
+
+        let more_messages = self.read_remaining_messages(msg_length)?;
+
+        let msg = self.extract_msg(msg_length)?;
+        tracing::debug!(target: "network", "received PeerMessage::{} len: {}", msg, msg_length);
+        messages.push((msg, first_byte_time));
+
+        for (len, timestamp) in more_messages {
+            let msg = self.extract_msg(len)?;
+            tracing::debug!(target: "network", "received PeerMessage::{} len: {}", msg, len);
+            messages.push((msg, timestamp));
+        }
+
+        // make sure we can probably read the next message in one syscall next time
+        let max_len_after_next_read = self.buf.chunk_mut().len() + self.buf.remaining();
+        if max_len_after_next_read < 512 {
+            self.buf.reserve(512 - max_len_after_next_read);
+        }
+        Ok(messages)
+    }
+
+    async fn recv_message(&mut self) -> io::Result<(PeerMessage, Instant)> {
+        let (msg_length, first_byte_time) = self.read_msg_length().await?;
+
+        while self.buf.remaining() < msg_length + 4 {
+            self.do_read().await?;
+        }
+        Ok((self.extract_msg(msg_length)?, first_byte_time))
+    }
+
+    /// Reads from the socket until we receive some message that we care to pass to the caller
+    /// (that is, represented in `Message`). After receiving the first such message, we will
+    /// continue to read any messages available to be read without blocking. The `Instant`s
+    /// in the returned `Vec` are timestamps taken when the corresponding messages were read
+    pub async fn recv(&mut self) -> io::Result<Vec<(Message, Instant)>> {
```

Comment:
Ok, I'm not exactly sure what's the goal, but it seems to me that this code is a bit overly complex. I've written sth like this to parse the message: https://github.com/near/nearcore/blob/06711f8460f946b8d2042aa1df6abe03c5184767/chain/network/src/peer/stream.rs#L155 and its essentially:
`
n = read.read_u32_le().await
let mut buf = vec![0; n];
read.read_exact(&mut buf[..]).await?;
`
where `read` is a TCP socket wrapped in a standard buffer of fixed size.

Reply:
made it simpler, although can't do `read_u32_le()` here, as then the future would not be cancellation-safe and not usable with `tokio::select`. actually did run into this a while ago when I was first working on this: https://github.com/near/nearcore/pull/7541/commits/ed5d20e5cc448e68988abb01f242f3f9fba30e53